rosbag2_cpp/src/rosbag2_cpp/writer.cpp

```diff
@@ -112,11 +114,46 @@ void Writer::write(
   serialized_bag_message->topic_name = topic_name;
   serialized_bag_message->time_stamp = time.nanoseconds();
 
-  // temporary store the payload in a shared_ptr.
-  // add custom no-op deleter to avoid deep copying data.
-  serialized_bag_message->serialized_data = std::shared_ptr<rcutils_uint8_array_t>(
-    const_cast<rcutils_uint8_array_t *>(&message.get_rcl_serialized_message()),
-    [](rcutils_uint8_array_t * /* data */) {});
+  if (writer_impl_->request_owned_serialized_data()) {
```

Comment:
While we are adding these workarounds, can you also mark this function signature as `[[deprecated()]]` in the header? Based on the need to duplicate every message like this, I think we want to discourage its use and remove it later

Reply:
But this interface is called now by below template interface  
https://github.com/ros2/rosbag2/blob/622b0ada5bf78da81a24e33f00c0ff89097c80c0/rosbag2_cpp/include/rosbag2_cpp/writer.hpp#L149-L170

So I think we also need to modify this template interface.  

I want to use this new interface to replace this interface to be deprecated.  
https://github.com/Barry-Xu-2018/rosbag2/blob/5f18ae59f1358ae045e1e4b14afadbc333a8737d/rosbag2_cpp/include/rosbag2_cpp/writer.hpp#L149-L166

New interface also can be called instead of below codes. 
https://github.com/ros2/rosbag2/blob/622b0ada5bf78da81a24e33f00c0ff89097c80c0/rosbag2_transport/src/rosbag2_transport/recorder.cpp#L222-L246

What do you think ?